pkg/projectrequestlimit/validation.go

```diff
@@ -0,0 +1,253 @@
+	userv1listers "github.com/openshift/client-go/user/listers/user/v1"
+	"github.com/openshift/project-request-limit/pkg/response"
+)
+
+const (
+	// allowedTerminatingProjects is the number of projects that are owned by a user, are in terminating state,
+	// and do not count towards the user's limit.
+	allowedTerminatingProjects = 2
+	defaultResyncPeriod        = 4 * time.Hour
+	Resource                   = "projectrequestlimits"
+	Singular                   = "projectrequestlimit"
+	Name                       = "projectrequestlimit"
+)
+
+type Validator interface {
+	Validate(request *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse
+}
+
+func NewValidator(kubeClientConfig *restclient.Config, stopCh <-chan struct{}) (Validator, error) {
+	client, err := kubernetes.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	userClient, err := userclient.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	projectClient, err := projectclient.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// ProjectRequestLimit shared informer
+	piFactory := projectinformers.NewSharedInformerFactory(projectClient, defaultResyncPeriod)
+	prlInformer := piFactory.Project().V1().ProjectRequestLimits().Informer()
+	go prlInformer.Run(stopCh)
+
+	// User shared informer
+	userFactory := userinformers.NewSharedInformerFactory(userClient, defaultResyncPeriod)
+	userInformer := userFactory.User().V1().Users().Informer()
+	go userInformer.Run(stopCh)
+
+	// Namespace shared informer
+	nsFactory := informers.NewSharedInformerFactory(client, defaultResyncPeriod)
+	nsInformer := nsFactory.Core().V1().Namespaces().Informer()
+	go nsInformer.Run(stopCh)
+
+	// Wait for informer caches to sync
+	syncResult := piFactory.WaitForCacheSync(stopCh)
+	if r, ok := syncResult[reflect.TypeOf(prlInformer)]; ok && !r {
+		return nil, fmt.Errorf("failed to wait for ProjectRequestLimit informer cache to sync")
+	}
+	syncResult = userFactory.WaitForCacheSync(stopCh)
+	if r, ok := syncResult[reflect.TypeOf(userInformer)]; ok && !r {
+		return nil, fmt.Errorf("failed to wait for User informer cache to sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, nsInformer.HasSynced) {
+		return nil, fmt.Errorf("failed to wait for Namespace informer cache to sync")
+	}
+
+	return &projectRequestLimitValidator{
+		prlLister:  piFactory.Project().V1().ProjectRequestLimits().Lister(),
+		userLister: userFactory.User().V1().Users().Lister(),
+		nsLister:   nsFactory.Core().V1().Namespaces().Lister(),
+	}, nil
+}
+
+type projectRequestLimitValidator struct {
+	prlLister  projectv1listers.ProjectRequestLimitLister
+	userLister userv1listers.UserLister
+	nsLister   corev1listers.NamespaceLister
+}
+
+var _ Validator = (*projectRequestLimitValidator)(nil)
+
+func (p *projectRequestLimitValidator) Validate(request *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse {
```

Comment:
Implementation from [3.11](https://github.com/openshift/openshift-apiserver/blob/6ae368b1f0170d8b96dcc926fe13a541ba6e3bd1/pkg/project/apiserver/admission/requestlimit/admission.go#L88)

Reply:
while you copy this around (which is :+1:), can we also take the tests? I'm afraid this is a bunch of logic with no safety net in case we need to change it in the future